ejb3/src/main/java/org/jboss/as/ejb3/util/MdbValidationsUtil.java

```diff
@@ -0,0 +1,52 @@
+package org.jboss.as.ejb3.util;
+
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.as.ejb3.logging.EjbLogger;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.MethodInfo;
+
+public final class MdbValidationsUtil {
+
+    private MdbValidationsUtil() {
+
+    }
+
+    /**
+     * Returns true if the passed <code>mdbClass</code> meets the requirements set by the EJB3 spec about bean implementation
+     * classes. The passed <code>mdbClass</code> must not be an interface and must be public and not final and not abstract. If
+     * it passes these requirements then this method returns true. Else it returns false.
+     *
+     * @param mdbClass The MDB class
+     * @return
+     */
+    public static boolean assertMDBClassValidity(final ClassInfo mdbClass) {
+        final String className = mdbClass.name().toString();
+        short flags = mdbClass.flags();
+        // must *not* be an interface
+        if (Modifier.isInterface(flags)) {
+            EjbLogger.DEPLOYMENT_LOGGER.mdbClassCannotBeAnInterface(className);
+            return false;
+        }
+        // bean class must be public, must *not* be abstract or final
+        if (!Modifier.isPublic(flags) || Modifier.isAbstract(flags) || Modifier.isFinal(flags)) {
+            EjbLogger.DEPLOYMENT_LOGGER.mdbClassMustBePublicNonAbstractNonFinal(className);
+            return false;
+        }
+        // bean class can not have final method
+        List<String> finalMethods = new ArrayList<>(0);
+        for (MethodInfo method : mdbClass.methods()) {
```

Comment:
This is not actually what the spec says (also as ClassInfo does not cover inheritance this check is not correct anyway):

5.6.4 Message Listener Method
The message-driven bean class must define the message listener methods. The signature of a message listener method must follow these rules:
The method must be declared as public.
The method must not be declared as final or static.

5.6.5 Message-Driven Bean with No-Methods Listener Interface
The following additional requirements apply for a message-driven bean with a no-methods interface:
• All non-static public methods of the bean class and of any superclasses except
java.lang.Object are exposed as message listener methods.
• Only private methods of the bean class and any superclasses except java.lang.Object may be declared final.

So basically this validation is stricter than what the spec says and can cause valid deployments to fail. It also does not check that the methods are public, and does not take superclass methods into account.